### PR TITLE
Add `Span::prepare_for_async` method and `AbstractSpan` trait.

### PR DESCRIPTION
**Pull request by jmjoy**
Sometimes we need to create a span following by the last active span instead of as the a child span, like in https://github.com/apache/skywalking-php/blob/eabb210b42ff652b8e34d8d99a933b5489b80397/src/plugin/plugin_curl.rs#LL276C9-L276C9

1. Add `Span::prepare_for_async` mehtod, which will return the `AsyncSpan` object, when it dropped, it will auto finished and notify `TracingContext`.
2. Add `TracingContext::wait` method, which will block to wait all `AsyncSpan` finished.
3. Add `AbstractSpan` trait, `Span` and `AsyncSpan` all implement it.

### CODE REVIEW

**Comment by wu-sheng** [comment]
By following, you mean sharing the  same parent span with the current active span? Two active spans created and running?

**Comment by jmjoy** [comment]
> By following, you mean sharing the same parent span with the current active span? Two active spans created and running?

Yes, like `curl_multi_*`, many HTTP requests are requested simultaneously in a multi-threaded manner, so they should belong to the same parent. Is there a better way?

**Comment by wu-sheng** [comment]
If they run concurrently, usually we prefer span created and change into async mode(not be active span anymore)

https://skywalking.apache.org/docs/skywalking-java/next/en/setup/service-agent/java-agent/java-plugin-development-guide/#async-span-apis

This is Java dev doc. Notice, Java trace segment is in one thread only, so, when thread crossing happens, you need to go for new segment. But then the span duration can't represent the HTTP call anymore. So, we need async span to create in thread A, but finish in thread B. This concept should fit your scenarios as well.

**Comment by wu-sheng** [comment]
follow up span was introduced by Opentracing, but very confusing. So, OTEL doesn't mention that, SkyWalking never have this concept too.

File: e2e/data/expected_context.yaml [hunk 1]
```diff
@@ -50,7 +50,7 @@ segmentItems:
       peer: consumer:8082
       skipAnalysis: false
       spanId: 1
-      spanLayer: Http
+      spanLayer: Unknown
```
Comment (wu-sheng) [inline e2e/data/expected_context.yaml:53]: From http to unknown? Why?

**Review by wu-sheng — COMMENTED** [review]
Please update doc about new APIs.

File: e2e/data/expected_context.yaml (hunk 1 above)
Comment (jmjoy) [reply e2e/data/expected_context.yaml:53]: Just change the default spanlayer of ExitSpan to Unknown, and users can set what they want through setLayer.

**Comment by jmjoy** [comment]
> Please update doc about new APIs.

The documentation for this part of the API has already been mentioned in the document comments (generated in docs.rs), and mentioning it on readme will make it very large?

File: e2e/data/expected_context.yaml (hunk 1 above)
Comment (wu-sheng) [reply e2e/data/expected_context.yaml:53]: I see.

**Comment by wu-sheng** [comment]
Yes, such as how to typically use these new APIs.

**Review by wu-sheng — APPROVED** [review]

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**